proposals/NNNN-outofproc-compiler-api-architecture.md

```diff
@@ -0,0 +1,271 @@
+(see [include handler support](#include-handler-support)).
+
+### JSON-RPC messages (generic)
+
+#### JSON Request
+```json
+{
+    "json-rpc":"2.0",
+    "method":"methodname",
+    "params":{ "arg":"value",  "arg2":"value2", "arg3":"value3" },
+    "id":1
+}
+```
+#### JSON Response (success) - contains result data
+```json
+{
+    "json-rpc":"2.0",
+    "result":{ "res":"value" },
+    "id":1
+}
+```
+#### JSON Response (error) - contains error code
+```json
+{
+    "jsonrpc": "2.0",
+     "error": {"code": -32601, "message": "Method not found"},
+     "id": 1
+}
+```
+
+### Protocol communication with example JSON-RPC messages
+A simple compilation with no include handler callback provided will be
+implemented as a single `Request` followed by a single `Response`.
+
+```
+Syntax for describing Request/Response pairs:
+--> data sent to server
+<-- data coming from server
+```
+
+```
+---> REQUEST  id:1 compile...
+<--- RESPONSE id:1 compile succeeded/failed
+```
+
+STDOUT and STDERR will be captured and sent back with every completed
+compilation `Response`.  This ensures that the caller can see any additional
+information about the compile operation.
+
+STDOUT/STDERR will be returned in one of two forms determined by the caller.
+* A path to a file that contains the captured output. The caller is responsible
+for cleaning up this file after being accessed.
+* A serialized object that contains the contents of the captured output.
+> TBD: Is the serialized object a compressed buffer of the full contents or
+a stream object where the caller can read from?
+
+### Example: compile, no include handler, success
+```
+---> REQUEST           id:1 compile...
+<--- RESPONSE (result) id:1 compile succeeded
+
+(RPC-JSON example)
+
+--> {"jsonrpc": "2.0", "method": "compile",
+     "params": {"arg1": 23, "arg2": 42}, "id": 1}
+
+<-- {"jsonrpc": "2.0", "result": {"stdout": "somepath", "stderr": "somepath"},
+     "id": 1}
+```
+
+### Example: compile, no include handler, failure
+```
+---> REQUEST          id:1 compile...
+<--- RESPONSE (error) id:1 compile failed
+
+(RPC-JSON example)
+
+--> {"jsonrpc": "2.0", "method": "compile",
+     "params": {"arg1": 23, "arg2": 42}, "id": 1}
+
+<-- {"jsonrpc": "2.0", "error": {"code": -2, "message": "compile failed"},
+     "data": {"stdout": "somepath", "stderr": "somepath"}, "id": 1}
+```
+
+### Include handler support
+Include handlers add some additional complexity but still keep to the
+`Request`/`Response` pair design.  If an include handler callback is provided
+to the c api, a special proxy implementation of that callback is created and 
+connected to the IPC messaging system and lives in the clang.exe process.
+This proxy callback will be used during compilation when an include is
+required. Calling the proxy callback results in an IPC communication with
+the client's dispatcher in the form of a `Request`/`Response` pair asking for a 
+buffer for the specified include.
```

Comment:
The current include handler architecture does a round-trip through the include handler for each potential include path constructed.

I think we will likely want to locate this include path search sequence on the client so it does not need to round trip through RPC for each search possibility.  That brings the round-trip frequency down from `includes * locations` to just `includes`.

That would mean capturing the path for the file containing the include in the "need include" request, then cycling through this and the locally tracked include paths in the client process and calling the include handler for each one until a match is found (or not), then issuing the response.

Reply:
I think there are a couple different reasons why custom include handlers are useful:
1) They identify include dependencies.
2) They allow remapping directory structures.
3) They allow treating in-memory objects as files.

For (1), we should discourage using include handlers as the solution and instead have users adopt the compiler-generated dep files.

For (2), clang has a poorly-documented "header maps" feature that we could use, which would radically simplify our implementation needs.

For (3), having the server process doing file IO, defeats the point, so we'll likely need some model where we can communicate buffers of data from the host process to the compiler process. Streaming the data across a named pipe might be an acceptable initial implementation, but we likely need to do some shared memory solution to get decent performance and avoid redundant copies.